channeld/channeld.c

```diff
@@ -2381,36 +2390,58 @@ static void peer_reconnect(struct peer *peer,
 	}
 }
 
-/* Funding has locked in, and reached depth. */
-static void handle_funding_locked(struct peer *peer, const u8 *msg)
+/* ignores the funding_depth unless depth >= minimum_depth
+ * (except to update billboard, and set peer->depth_togo). */
+static void handle_funding_depth(struct peer *peer, const u8 *msg)
 {
-	unsigned int depth;
+	u32 depth;
+	struct short_channel_id *scid;
 
-	if (!fromwire_channel_funding_locked(msg,
-					     &peer->short_channel_ids[LOCAL],
-					     &depth))
-		master_badmsg(WIRE_CHANNEL_FUNDING_LOCKED, msg);
+	if (!fromwire_channel_funding_depth(tmpctx,
+					    msg,
+					    &scid,
+					    &depth))
+		master_badmsg(WIRE_CHANNEL_FUNDING_DEPTH, msg);
 
 	/* Too late, we're shutting down! */
 	if (peer->shutdown_sent[LOCAL])
 		return;
 
-	if (!peer->funding_locked[LOCAL]) {
-		status_trace("funding_locked: sending commit index %"PRIu64": %s",
-			     peer->next_index[LOCAL],
-			     type_to_string(tmpctx, struct pubkey,
-					    &peer->next_local_per_commit));
-		msg = towire_funding_locked(NULL,
-					    &peer->channel_id,
-					    &peer->next_local_per_commit);
-		sync_crypto_write(&peer->cs, PEER_FD, take(msg));
-		peer->funding_locked[LOCAL] = true;
-	}
+	if (depth < peer->channel->minimum_depth) {
+		if (peer->funding_locked[LOCAL] || peer->depth_togo == 0)
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
```

Comment:
This error made me stumble a bit. Are we sure this can't happen during restarts with rescans?

Reply:
Hi, @cdecker I'm not sure if what I think is correct. Hope you don't mind my pool English.
 
In the original code, the channel will be saved in the DB in this [code block](https://github.com/ElementsProject/lightning/blob/0d35a7111a10af070bb6f9aaf53b6a27ca38b264/lightningd/peer_control.c#L875).
But now, to prevent saving channel with unlocked funding, I add the condition `local_locked && !channel->scid`. So, these channel with unlocked funding won't be saved until depth >= minimum_depth.

And the struct peer in Channeld always inits from [there](https://github.com/ElementsProject/lightning/blob/0d35a7111a10af070bb6f9aaf53b6a27ca38b264/channeld/channeld.c#L2902). The peer in Channeld won't be saved in DB and it is only dependent on the channel information in the Lightningd.

I suppose when we restart, we need rebuild this peer struct(in Channeld) according to the channel(in Lightningd). If the funding locked before, there may be 4 situations:
1. the depth >= minimum_depth(we can get the topo from DB) , and the channel has been saved in DB before: we can init the peer with `peer->funding_locked[LOCAL] == true`([there](https://github.com/ElementsProject/lightning/blob/0d35a7111a10af070bb6f9aaf53b6a27ca38b264/channeld/channeld.c#L2703) we use if `channel->scid` exist to init `funding_locked[LOCAL]`), and `peer->depth_togo == minimum_depth`(I directly set this value to peer->depth_togo). But after some minutes, Lightningd will tell us the depth >= minimum_depth, we won't meet this error.
2. the depth >= minimum_depth, but we didn't save the channel in time: we can init the peer with `peer->funding_locked[LOCAL] = false`, and we alse init the `peer->depth_togo = minimum_depth` directly. 
3. the depth < minimum_depth, and we saved the channel in the DB: it is impossible! we only save the channel under the condition that funding is locked.
4. the depth < minimum_depth, and we didn't save the channel: we can init peer with `funding_locked[LOCAL] = false` and `peer->depth_togo = minimum_depth`.

So I think we won't meet this error when restart. What do you think about this?

Reply:
Hi!@cdecker I think this error won't happen unless the topo reorganize with setting too samll minimum_depth.
The peer we used in Channeld is really initialed according the channel_init message from Lightning. And we never save the peer in the DB directly. In other words, during every start, we need initial the peer->funding_locked[LOCAL] from Lightningd(peer->funding_locked[LOCAL] corresponds if channel->scid exist). So the 4 situations I listed above are reasonable.

I think the only situation that this error happens is:
Lightningd find the funding depth changes and don't hit minimum_depth, then Lightningd tell us the depth now. But before that, funding has hit minimum_depth once and we has drived the channel->scid. The reason of this situation maybe the minimum_depth is too small and topo reorganized.
We don't limit the smallist minimum_depth(BOLT2 just ask mimimum_depth can't be too big, and clightning ask it must be smaller than 10). But I think we don't need to set the smallest minimum_depth. This condition here can be a kind of alert that reorg has happend.
what do you think about?

Reply:
I know this goes off-topic, but currently we [allow](https://github.com/ElementsProject/lightning/blob/e843e138fe43c48c6165d860ea99a627945a91a4/lightningd/options.c#L606) small `minimum_depth = 1` (the [default](https://github.com/ElementsProject/lightning/blob/e843e138fe43c48c6165d860ea99a627945a91a4/lightningd/options.c#L543) is 3), which is easily reorged and can change `scid`. I actually think that `scid` should not be calculated before `ANNOUNCE_MIN_DEPTH` (=6), do we really need the `scid` before that?

Reply:
So IMHO the above check can be removed.

Reply:
@SimonVrouwe You'are right, and I should delete these checks here.
And I also notice a extreme case: 
Suppose we set a private channel (don't set announcement flag), and set a too small minimum locking depth (as you say, 1). When we notice the funding tx locked and we drive `short_channel_id` and store into the DB, but unfortunately, our node crash. When we restart, the reorg happens.
Now we initial our channel from DB with old `short_channel_id`, and we may can't be told that reorg happened because we delete the corresponding [`watchtx`](https://github.com/ElementsProject/lightning/blob/c7ab5102ba68ef90209bb4cfc0ff1e63fa68598a/lightningd/peer_control.c#L907). 
The next step the short_channel_id will be use is update_htlc_add (there need a check?).

Reply:
> ... and we may can't be told that reorg happened because we delete the corresponding watchtx.

Good catch!
Wallet transactions are never deleted from the db, but at reorg their `height` field is [set](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/wallet/db.c#L311) to NULL. When a block is reorged-out, all tx's with that height are [selected](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/lightningd/chaintopology.c#L658), and their `cb` is [fired](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/lightningd/chaintopology.c#L663) with depth 0. 

But if the `watch` was [deleted](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/lightningd/peer_control.c#L908), after only 1 confirm when `minimum_depth=1` and ANNOUNCE_FLAG unset, the [watch](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/lightningd/watch.c#L140) is gone and [nothing](https://github.com/ElementsProject/lightning/blob/2af3996bea21c91945e621aa49fe461fe3e39a0e/lightningd/watch.c#L247) happens!

@cdecker So this looks like a bug? Although it is a _very_ rare case (how often do tx get reorged to another height?), I will make a new issue about this.

Reply:
It's a really rare case. The main problem I think is too samll minimum_depth. 
Maybe we can set limit of the minimum_depth, or any other operations to handle it(fail the channel, or generate warning...).